canisters/src/vts/src/lib.rs

```diff
@@ -47,12 +85,61 @@ fn request_firmware() -> VTSResult<()> {
     Ok(())
 }
 
+#[derive(CandidType, Deserialize, Debug)]
+pub struct UploadFirmwareRequest {
+    pub principal: String,
+    pub _firmware: Vec<u8>,
+    pub _arch: String,
+}
+
 #[ic_cdk::update]
 fn upload_firmware(req: UploadFirmwareRequest) -> VTSResult<()> {
     // todo: this canister method should be executed only by our gateway.
     FIRMWARE_REQUESTS.with(|requests| requests.borrow_mut().remove(&req.principal));
     Ok(())
 }
 
+#[ic_cdk::update]
+fn create_agreement(
+    name: String,
+    vh_customer: Principal,
+    daily_usage_fee: String,
+    gas_price: String,
+) -> VTSResult<u128> {
+    let caller = ic_cdk::api::caller();
+    ic_cdk::println!("requested agreement creation by {}", caller);
+    AGREEMENTS.with(|agreements| {
+        let mut agreements = agreements.borrow_mut();
+
+        if agreements.iter().any(|(_, agreement)| {
+            agreement.name == name && agreement.vh_customer == vh_customer &&
+            agreement.conditions.daily_usage_fee == daily_usage_fee &&
+            agreement.conditions.gas_price == gas_price
+        }) {
+            return Err(Error::AlreadyExists);
+        }
+
+        let mut next_agreement_id = agreements.len() as u128;
```

Comment:
It is better to store counter as u128 in canister state. Because .len return usize, and usize maximum value is u32/u64.